app/lib/metrics_service.rb

```diff
@@ -124,4 +124,65 @@ def self.log(name, props={}, user)
     self.post_to_bard(params, user)
 
   end
+
+  # Log error metrics to Mixpanel via Bard web service
+  # Handled from rescue_from blocks in controllers
+  #
+  # Bard docs:
+  # https://terra-bard-prod.appspot.com/docs/
+  #
+  # @param {Exception} exception - instance of exception that was caught
+  # @param {ActionDispatch::Request} request - request object in which exception was thrown
+  # @param {User} user - User model object, if present
+  # @param {Study} study - Study model object, if present
+  def self.report_error(exception, request, user=nil, study=nil)
+    Rails.logger.error "Reporting error analytics to mixpanel for (#{exception.class.name}) #{exception.message}"
+
+    # error properties hash details
+    # appPath: request URL where error occurred
+    # serverMethod: Rails controller & method that corresponds to the error
+    # type: error Ruby class
+    # text: error message/details
+    # appId: identifier for Single Cell Portal
+    # env: Rails environment
+    # studyAccession: corresponding SCP study (if present)
+    # authenticated: user auth status
+    # registeredForTerra: status of user Terra registration (if present)
+    # distinct_id: user metrics UUID for identifying users in Bard
+    props = {
+      appPath: request.fullpath,
```

Comment:
Blocker: `appPath` must omit study name, which exists in some URLs.  Ideally,`appPath` logged from server should align with `appPath` logged from client, to enable easy tracing between client and server event. (Solving the former would aid the latter.)

See [`appPath` and `appFullPath` usage](https://github.com/broadinstitute/single_cell_portal_core/blob/e7882646087c6eed0d9023b4d9dbff150815ee09/app/javascript/lib/metrics-api.js#L267), and relevant function definitions in [`getAnalyticsPageName`](https://github.com/broadinstitute/single_cell_portal_core/blob/e7882646087c6eed0d9023b4d9dbff150815ee09/app/javascript/lib/metrics-api.js#L247) and [`getAppFullPath`](https://github.com/broadinstitute/single_cell_portal_core/blob/e7882646087c6eed0d9023b4d9dbff150815ee09/app/javascript/lib/metrics-api.js#L238).

This special handling exists because workspace name (and, analogously, study name) should not be logged to Mixpanel per DSP policy; see requirement 1 [here](https://docs.google.com/document/d/1di8uwv-nDMNJp83Gs9Q_mlZp3q1DFboXGPLzbsbPyn8/edit#heading=h.n7z2sg51xcwg).  For example, if an error occurred in https://singlecell.broadinstitute.org/single_cell/study/SCP555/personally-identifiable-information-in-title, that special handling prevents the sensitive information in the title from being logged to Mixpanel.

Reply:
849ddd9